gms/gossiper.cc

```diff
@@ -224,13 +224,61 @@ future<> gossiper::handle_syn_msg(msg_addr from, gossip_digest_syn syn_msg) {
         return make_ready_future<>();
     }
 
+    syn_msg_pending& p = _syn_handlers[from.addr];
+    if (p.pending) {
+        // The latest syn message from peer has the latest infomation, so
+        // it is safe to drop the previous syn message and keep the latest
+        // one only.
+        logger.debug("Queue gossip syn msg from node {}, syn_msg={}", from, syn_msg);
+        p.syn_msg = std::move(syn_msg);
+        return make_ready_future<>();
+    } else {
+        // Process the syn message immediately
+        logger.debug("Process gossip syn msg from node {}, syn_msg={}", from, syn_msg);
+        p.pending = true;
+        return do_with(std::move(syn_msg), [this, from] (gossip_digest_syn& syn_msg) mutable {
+            return repeat([this, from, &syn_msg] {
+                return do_send_ack_msg(from, std::move(syn_msg)).then([this, from, &syn_msg] () mutable {
+                    if (!_syn_handlers.count(from.addr)) {
```

Comment:
What prevents the gossiper object from getting destroyed before this executes?

Reply:
Gossiper object is seastar::async_sharded_service, we could take a ref g = this->shared_from_this().  Updated in new version.